### DISCUSSION TITLE
How to add custom dictionary?

### DISCUSSION THREAD

**Discussion by yangqinj** [explosion/spaCy#6606] [Help: Coding & Implementations] [answered]
## Your Environment

- Operating System: centos
- Python Version Used: 3.8.2
- spaCy Version Used: 2.3.5
- Environment Information:

Hi, I am trying use spaCy for tokenizing for Chinese, English and Spanish. The result seems great! But I cannot find document details how to add custom dictionary to prevent two words from seperating. For example, I hope "New York" to be a word for English, "北京大学" to be a word for Chinese. Thanks!

**Comment by svlandeg** [comment]
Hi!

FYI - we converted this question to a thread on the discussion board.

We're aiming to keep the Issue tracker focused mainly on bug reports, while the new GH Discussion board can be a place where the community can come together, help eachother, discuss best practices, and so on.

**Comment by polm** [comment]
It sounds like you want a [tokenizer exception](https://spacy.io/api/tokenizer#add_special_case). From the docs:

```
from spacy.attrs import ORTH, NORM
case = [{ORTH: "do"}, {ORTH: "n't", NORM: "not"}]
tokenizer.add_special_case("don't", case)
```

**Comment by adrianeboyd** [comment, marked as answer]
Tokenizer exceptions will depend on the underlying tokenizer. For the `Tokenizer` used for most languages that don't require separate word segmentation tools (exceptions are Chinese, Japanese, Korean, Thai, and Vietnamese), exceptions with spaces in them aren't supported in v2, but will work in v3. The functionality is implemented, but there's a bug in 3.0.0rc2 that should be fixed in 3.0.0rc3 by PR #6668.

In spacy v2, you could implement a custom component that finds the phrases to merge and uses the retokenizer to merge tokens like for `New York`. You can use the `PhraseMatcher` to identify and have a look at the `merge_entities` pipeline component to see an example of how to do this with the retokenizer.

For the Chinese pkuseg tokenizer, you can modify the user dictionary as described in the expandable pkuseg API section here: https://spacy.io/usage/models#chinese . I think it's possible with user dictionaries for jieba, too, but the current `ChineseTokenizer` API doesn't include any methods for modifying these settings for jieba.
